include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -445,47 +450,140 @@ auto applyDeflation(const SizeType i_begin, const SizeType i_end, RhoSender&& rh
          ex::ensure_started();
 }
 
+// z is an input whose values are destroyed by this call (input + workspace)
 template <class T, class KSender, class RhoSender>
 void solveRank1Problem(const SizeType i_begin, const SizeType i_end, KSender&& k, RhoSender&& rho,
-                       Matrix<const T, Device::CPU>& d, Matrix<const T, Device::CPU>& z,
+                       Matrix<const T, Device::CPU>& d, Matrix<T, Device::CPU>& z,
                        Matrix<T, Device::CPU>& evals, Matrix<T, Device::CPU>& evecs) {
   namespace ex = pika::execution::experimental;
   namespace di = dlaf::internal;
 
   const SizeType n = problemSize(i_begin, i_end, evals.distribution());
   const SizeType nb = evals.distribution().blockSize().rows();
 
-  auto rank1_fn = [n, nb](auto k, auto rho, auto d_tiles_futs, auto z_tiles_futs, auto eval_tiles,
-                          auto evec_tiles) {
-    const TileElementIndex zero(0, 0);
-    const T* d_ptr = d_tiles_futs[0].get().ptr(zero);
-    const T* z_ptr = z_tiles_futs[0].get().ptr(zero);
-    T* eval_ptr = eval_tiles[0].ptr(zero);
+  TileCollector tc{i_begin, i_end};
 
-    matrix::Distribution distr(LocalElementSize(n, n), TileElementSize(nb, nb));
+  const std::size_t nthreads = getTridiagRank1NWorkers();
+  ex::start_detached(
+      ex::when_all(ex::just(std::make_shared<pika::barrier<>>(nthreads)), std::forward<KSender>(k),
+                   std::forward<RhoSender>(rho), ex::when_all_vector(tc.read(d)),
+                   ex::when_all_vector(tc.readwrite(z)), ex::when_all_vector(tc.readwrite(evals)),
+                   ex::when_all_vector(tc.readwrite(evecs)),
+                   ex::just(std::vector<memory::MemoryView<T, Device::CPU>>())) |
+      ex::transfer(di::getBackendScheduler<Backend::MC>(pika::execution::thread_priority::high)) |
+      ex::bulk(nthreads, [nthreads, n, nb](std::size_t thread_idx, auto& barrier_ptr, auto& k, auto& rho,
+                                           auto& d_tiles_futs, auto& z_tiles, auto& eval_tiles,
+                                           auto& evec_tiles, auto& ws_vecs) {
+        common::internal::SingleThreadedBlasScope single;
```

Comment:
I realize now that the scope of this has to be narrower. There's a risk that the barrier yields and the pika task is resumed on another threads, in which case the destructor of `SingleThreadedBlasScope` may reset the number of threads on the wrong worker thread.

I will check if this also needs to be changed in the other bulk algorithms. We can also have a debug check to make sure the same worker thread sets and restores the number of threads.

Reply:
If I get it correctly, we should not end up having `SingleThreadedBlasScope` crossing the boundary of a barrier yield.

If that's correct, I will fix it accordingly in this bulk, where we will need two of them, one for `laed4` and one for `nrm2`.

Reply:
Correct.

More generally: anything that may yield should not have a `SingleThreadedBlasScope` around it (the number of places where this can happen "silently" will slightly reduce with pika 0.16.0 because of https://github.com/pika-org/pika/pull/672).

Reply:
Fixed in https://github.com/eth-cscs/DLA-Future/pull/860/commits/c31139be7001630d8cc9d0af6548252fc0971198

Reply:
Perfect, thanks.